src/main/java/io/cdap/plugin/dlp/Utils.java

```diff
@@ -0,0 +1,217 @@
+
+    if (table.getRowsCount() == 0) {
+      throw new Exception("DLP returned a table with no rows");
+    }
+    Table.Row row = table.getRows(0);
+    for (int i = 0; i < table.getHeadersList().size(); i++) {
+
+      String fieldName = table.getHeadersList().get(i).getName();
+
+      Value fieldValue = row.getValues(i);
+      Schema tempSchema = oldRecord.getSchema().getField(fieldName).getSchema();
+      Schema fieldSchema = tempSchema.isNullable() ? tempSchema.getNonNullable() : tempSchema;
+      if (fieldValue == null) {
+        continue;
+      }
+
+      final Schema.LogicalType logicalType = fieldSchema.getLogicalType();
+      if (logicalType != null) {
+        switch (logicalType) {
+          case TIME_MICROS:
+          case TIME_MILLIS:
+
+            TimeOfDay timeValue = fieldValue.getTimeValue();
+            recordBuilder.setTime(fieldName, LocalTime
+              .of(timeValue.getHours(), timeValue.getMinutes(), timeValue.getSeconds(), timeValue.getNanos()));
+            break;
+
+          case TIMESTAMP_MICROS:
+          case TIMESTAMP_MILLIS:
+            Timestamp timestampValue = fieldValue.getTimestampValue();
+            ZoneId zoneId = oldRecord.getTimestamp(fieldName).getZone();
+            LocalDateTime localDateTime;
+            if (timestampValue.getSeconds() + timestampValue.getNanos() == 0) {
+              localDateTime = LocalDateTime
+                .parse(fieldValue.getStringValue(), DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm'Z'"));
+
+            } else {
+              localDateTime = Instant
+                .ofEpochSecond(timestampValue.getSeconds(), timestampValue.getNanos())
+                .atZone(zoneId)
+                .toLocalDateTime();
+            }
+
+            recordBuilder.setTimestamp(fieldName, ZonedDateTime.of(localDateTime, zoneId));
+            break;
+          case DATE:
+            Date dateValue = fieldValue.getDateValue();
+            recordBuilder
+              .setDate(fieldName, LocalDate.of(dateValue.getYear(), dateValue.getMonth(), dateValue.getDay()));
+            break;
+          default:
+            throw new IllegalArgumentException("Failed to parse table into structured record");
+
+        }
+
+
+      } else {
+        recordBuilder.convertAndSet(fieldName, fieldValue.getStringValue());
+      }
+    }
+    return recordBuilder.build();
+  }
+
+  public static StructuredRecord.Builder createBuilderFromStructuredRecord(StructuredRecord record) {
+    StructuredRecord.Builder recordBuilder = StructuredRecord.builder(record.getSchema());
+    for (Schema.Field field : record.getSchema().getFields()) {
+      String fieldName = field.getName();
+      Object fieldValue = record.get(fieldName);
+      Schema fieldSchema = field.getSchema().isNullable() ? field.getSchema().getNonNullable() : field.getSchema();
+
+      final Schema.LogicalType logicalType = fieldSchema.getLogicalType();
+      if (fieldSchema.getType().isSimpleType()) {
```

Comment:
dlp does not support complex type?

Reply:
No, it only supports simple types natively. We can try adding support for complex types from our end in a future version